File: auto/bin/parse_xml.pl
#!/usr/bin/perl

use strict;
use XML::LibXML;

sub findChildren
{
	my @result = ();

	my $len = @{$_[1]};
	my $nodeName = ${$_[1]}[0];

	if (1 == $len)
	{
		foreach my $i ($_[0]->getChildNodes())
		{
			if (XML_ELEMENT_NODE == $i->nodeType())
			{
				if ($nodeName eq $i->nodeName())
				{
					push @result, $i;
				}
			}
		}
	}
	else
	{
		$len--;
		my @sub_path = @{$_[1]}[1..$len];

		foreach my $i ($_[0]->getChildNodes())
		{
			if (XML_ELEMENT_NODE == $i->nodeType())
			{
				if ($nodeName eq $i->nodeName())
				{
					my @sub_result = findChildren($i, \@sub_path);
					my $pos = @result;
					splice @result, $pos, 0, @sub_result;
				}
			}
		}
	}
	return @result;
}

sub findData
{
	my @result = ();
	my @nodes = findChildren($_[0], \@{$_[1]});

	foreach my $i (@nodes)
	{
		push @result, $i->textContent();
	}

	return @result;
}

sub findParams
{
	my @name = ('name');
	my @nodes = findData($_[0], \@name);

	my $n = $nodes[0];
	my $t = '';

	foreach my $i ($_[0]->getChildNodes())
	{
		if (XML_TEXT_NODE == $i->nodeType())
		{
			$t .= $i->textContent();
		}

		if ((XML_ELEMENT_NODE == $i->nodeType()) and ('ptype' eq $i->nodeName()))
		{
			$t .= $i->firstChild()->textContent();
		}
	}

	my %ret = ();
	$ret{$t} = $n;
	return %ret;
}

sub findEnums
{
	my @enum = ('registry', 'enums', 'enum');
	my @nodes = findChildren($_[0], \@enum);
	my %ret = ();

	foreach my $i (@nodes)
	{
		my $n = $i->getAttribute('name');
		my $v = $i->getAttribute('value');
		$ret{$n} = $v;
	}

	return %ret;
}

sub findCommands
{
	my @command = ('registry', 'commands', 'command');
	my @proto = ('proto');
	my @param = ('param');
	my @nodes = findChildren($_[0], \@command);
	my %ret = ();

	foreach my $i (@nodes)
	{
		my @sub_nodes = findChildren($i, \@proto);
		my %r_n = findParams($sub_nodes[0]);

		my @p = ();
		@sub_nodes = findChildren($i, \@param);

		foreach my $j (@sub_nodes)
		{
			my %params = findParams($j);
			push @p, \%params;
		}

		@sub_nodes = keys %r_n;
		my $r = $sub_nodes[0];

		@sub_nodes = values %r_n;
		my $n = $sub_nodes[0];

		$ret{$n} = {$r, \@p};
	}

	return %ret;
}

sub findFeaturesExtensions
{
	my @enum = ('require', 'enum');
	my @command = ('require', 'command');
	my @nodes = findChildren($_[0], \@{$_[1]});
	my %ret = ();

	foreach my $i (@nodes)
	{
		my $n = $i->getAttribute('name');
		my @e = ();
		my @c = ();

		my @sub_nodes = findChildren($i, \@enum);

		foreach my $j (@sub_nodes)
		{
			my $value = $j->getAttribute('name');
			push @e, $value;
		}

		@sub_nodes = findChildren($i, \@command);

		foreach my $j (@sub_nodes)
		{
			my $value = $j->getAttribute('name');
			push @c, $value;
		}

		my @sub_ret = (\@e, \@c);
		$ret{$n} = \@sub_ret;
	}

	return %ret;
}

sub findFeatures
{
	my @feature = ('registry', 'feature');
	return findFeaturesExtensions($_[0], \@feature);
}

sub findExtensions
{
	my @extension = ('registry', 'extensions', 'extension');
	return findFeaturesExtensions($_[0], \@extension);
}

sub findApi
{
	my $dom = $_[0];
	my %enums = findEnums($dom);
	my %commands = findCommands($dom);
	my %features = findFeatures($dom);
	my %extensions = findExtensions($dom);

	return (\%enums, \%commands, \%features, \%extensions);
}

sub writeExtension
{
	my $f = $_[0];
	my $name = $_[1];
	my @extensions = @{$_[2]};
	my %enums = %{$_[3]};
	my %commands = %{$_[4]};

	print {$f} "$name\n";
	print {$f} "https://www.khronos.org/registry/egl/specs/eglspec.1.5.pdf\n";

	if ($name =~ m/_VERSION_/)
	{
		print {$f} "\n";
	}
	else
	{
		print {$f} "$name\n";
	}

	print {$f} "\n";
	my %enums_ = ();

	foreach my $j (@extensions)
	{
		my @arr = @{$j};
		my $j_ = $arr[0];
		@arr = @{$j_};

		foreach my $j__ (@arr)
		{
			my $enum = $enums{$j__};
			$enums_{$j__} = $enum;
		}
	}

	my %added_enums_ = ();

	foreach my $j (sort values %enums_)
	{
		my $v = $enums_{$j};

		foreach my $j_ (keys %enums_)
		{
			if ((not $added_enums_{$j_}) and ($j eq $enums{$j_}))
			{
				$added_enums_{$j_} = $j;
				print {$f} "\t$j_ $j\n";
			}
		}
	}

	my %commands_ = ();

	foreach my $j (@extensions)
	{
		my @arr = @{$j};
		my $j_ = $arr[1];
		@arr = @{$j_};

		foreach my $j__ (@arr)
		{
			my $command = $commands{$j__};
			$commands_{$j__} = $command;
		}
	}

	foreach my $j (sort keys %commands_)
	{
		my $value = $commands_{$j};
		my %value_ = %{$value};
		my @return_value = keys %value_;
		my $return_value_ = $return_value[0];

		my @params = values %value_;
		my $params_ = $params[0];
		my @params__ = @{$params_};

		my $str_params = '';

		foreach my $j_ (@params__)
		{
			my @types = keys %{$j_};
			my @names = values %{$j_};

			my $type = $types[0];
			my $name = $names[0];

			$str_params .= $type . ' ' . $name . ', ';
		}

		if (0 == length($str_params))
		{
			$str_params = ' void ';
		}
		else
		{
			$str_params = substr($str_params, 0, -2);
		}

		my $full_command = "\t$return_value_ $j ($str_params)\n";
		print ${f} "$full_command";
	}

	return 1;
}

if (@ARGV)
{
	my $filename = '';
	my $options_core = '';
	my $options_extensions = '';
	my $fill_core = 0;
	my $fill_extensions = 0;

	foreach my $argument (@ARGV)
	{
		if ($argument =~  m/^--core/)
		{
			$fill_core = 1;
		}
		elsif ($argument =~  m/^--extensions/)
		{
			$fill_extensions = 1;
		}
		elsif (0 == length($options_core) and 0 != $fill_core)
		{
			$options_core = $argument;
			$fill_core = 0;
		}
		elsif (0 == length($options_extensions) and 0 != $fill_extensions)
		{
			$options_extensions = $argument;
			$fill_extensions = 0;
		}
		elsif (0 == length($filename))
		{
			$filename = $argument;
		}
	}

	my $dom = XML::LibXML->load_xml(location => $filename);
	my @api = findApi($dom);

	my $enums_count = keys %{$api[0]};
	my $commands_count = keys %{$api[1]};
	my $features_count = keys %{$api[2]};
	my $extensions_count = keys %{$api[3]};

	print("Found $enums_count enums, $commands_count commands, $features_count features and $extensions_count extensions.\n");

	my %enums = %{$api[0]};
	my %commands = %{$api[1]};

	if (0 != length($options_core))
	{
		foreach my $feature (sort keys %{$api[2]})
		{
			my %values_ = %{$api[2]};
			my @values__ = $values_{$feature};
			my $file_path = $options_core . '/' . $feature;
			open(my $f, '>', $file_path); # or continue;
			writeExtension($f, $feature, \@values__, \%enums, \%commands);
			close($f);
			# print("$file_path\n");
		}
	}

	if (0 != length($options_extensions))
	{
		foreach my $ext (sort keys %{$api[3]})
		{
			my %values_ = %{$api[3]};
			my @values__ = $values_{$ext};
			my $file_path = $options_extensions . '/' . $ext;
			open(my $f, '>', $file_path); # or continue;
			writeExtension($f, $ext, \@values__, \%enums, \%commands);
			close($f);
			# print("$file_path\n");
		}
	}
}

File: auto/bin/parse_xml.py
#!/usr/bin/env python

import re
import sys
from xml.dom.minidom import parse, Node

#
# DOM traversal utility functions
#

def findChildren(node, path):
    result = []
    if len(path)==1:
        for i in node.childNodes:
            if i.nodeType==Node.ELEMENT_NODE:
                if i.tagName==path[0]:
                    result.append(i)
    else:
        for i in node.childNodes:
            if i.nodeType==Node.ELEMENT_NODE:
                if i.tagName==path[0]:
                    result.extend(findChildren(i, path[1:]))
    return result

def findData(node, path):
    return [ i.firstChild.data for i in findChildren(node, path) ]

isPointer = re.compile('(.*)([ ]+)([*]+)')

def findParams(node):
    n = findData(node, ['name'])[0]
    t = ''
    for i in node.childNodes:
        if i.nodeType==Node.TEXT_NODE:
            t += i.data
        if i.nodeType==Node.ELEMENT_NODE and i.tagName=='ptype':
            t += i.firstChild.data

    t.strip()
    m = isPointer.match(t)
    if m:
        t = ('%s%s'%(m.group(1), m.group(3))).strip()
    return ( t, n.strip())

def findEnums(dom):
    ret = {}
    for i in findChildren(dom, [ 'registry', 'enums', 'enum' ]):
      n = i.getAttribute('name')
      v = i.getAttribute('value')
      ret[n] = v
    return ret

def findCommands(dom):
    ret = {}
    for i in findChildren(dom, [ 'registry', 'commands', 'command' ]):
      r,n = findParams(findChildren(i, ['proto'])[0])
      p = [ findParams(j) for j in findChildren(i, ['param'])]
      ret[n] = (r, p)
    return ret

def findFeatures(dom):
    ret = {}
    for i in findChildren(dom, [ 'registry', 'feature' ]):
        n = i.getAttribute('name')
        e = []
        c = []
        for j in findChildren(i, [ 'require', 'enum' ]):
            e.append(j.getAttribute("name"))
        for j in findChildren(i, [ 'require', 'command' ]):
            c.append(j.getAttribute("name"))
        ret[n] = (e,c)
    return ret

def findExtensions(dom):
    ret = {}
    for i in findChildren(dom, [ 'registry', 'extensions', 'extension' ]):
        n = i.getAttribute('name')
        e = []
        c = []
        for j in findChildren(i, [ 'require', 'enum' ]):
            e.append(j.getAttribute("name"))
        for j in findChildren(i, [ 'require', 'command' ]):
            c.append(j.getAttribute("name"))
        ret[n] = (e,c)
    return ret

def findApi(dom, name):
    enums      = findEnums(dom)
    commands   = findCommands(dom)
    features   = findFeatures(dom)
    extensions = findExtensions(dom)
    return (enums, commands, features, extensions)

#
#
#

isWGL = re.compile('WGL_([A-Z0-9]+)_.*')

def writeExtension(f, name, extension, enums, commands):
    f.write(('%s\n'%name).encode())

    url = 'https://www.khronos.org/registry/egl/specs/eglspec.1.5.pdf'

    m = isWGL.match(name)
    if m:
        url = 'https://www.khronos.org/registry/OpenGL/extensions/%s/%s.txt'%(m.group(1), name)

    f.write(('%s\n'%(url)).encode())

    if name.find('_VERSION_')==-1:
        f.write(('%s\n'%name).encode())
    else:
        f.write('\n'.encode())
    f.write('\n'.encode())

    enums = [ (j, enums[j]) for j in extension[0] ]
    for e in sorted(enums, key=lambda i: i[1]):
        f.write(('\t%s %s\n'%(e[0], e[1])).encode())

    commands = [ (j, commands[j]) for j in extension[1] ]
    for c in sorted(commands):
        params = ', '.join( [ '%s %s'%(j[0].strip(), j[1].strip()) for j in c[1][1] ] )
        if len(params)==0:
            params = 'void'
        f.write(('\t%s %s (%s)\n'%(c[1][0].strip(), c[0].strip(), params)).encode())

if __name__ == '__main__':

  from argparse import ArgumentParser
  import os

  parser = ArgumentParser(description='usage: %prog [options] [XML specs...]')
  parser.add_argument("--core", dest="core", help="location for core outputs", default='')
  parser.add_argument("--api", dest="name", help="API name: egl, wgl, glx, etc", default='')
  parser.add_argument("--extensions", dest="extensions", help="location for extensions outputs", default='')

  (options, args) = parser.parse_known_args()
  options = vars(options)

  for i in args:

    dom = parse(i)
    api = findApi(dom, options['name'])

    print('Found {} enums, {} commands, {} features and {} extensions.'.format(
        len(api[0]), len(api[1]), len(api[2]), len(api[3])))

    if len(options['core']):
        for i in api[2].keys():
            f = open('%s/%s'%(options['core'], i), 'wb')
            writeExtension(f, i, api[2][i], api[0], api[1])
            f.close()

    if len(options['extensions']):
        for i in api[3].keys():
            f = open('%s/%s'%(options['extensions'], i), 'wb')
            writeExtension(f, i, api[3][i], api[0], api[1])
            f.close()

